daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala

```diff
@@ -0,0 +1,342 @@
+import scala.collection.compat.immutable.LazyList
+
+private final class Validation[Nid, Cid](implicit ECid: Equal[Cid]) {
+
+  import scalaz.std.option._
+  import scalaz.syntax.equal._
+
+  private[this] def keyIsReplayedBy(
+      recorded: Option[KeyWithMaintainers[Value[Cid]]],
+      replayed: Option[KeyWithMaintainers[Value[Cid]]],
+  ): Boolean = {
+    (recorded, replayed) match {
+      case (None, None) => true
+      case (Some(recordedValue), Some(replayedValue)) =>
+        keyIsReplayedBy(recordedValue, replayedValue)
+      case _ => false
+    }
+  }
+
+  private[this] def keyIsReplayedBy(
+      recorded: KeyWithMaintainers[Value[Cid]],
+      replayed: KeyWithMaintainers[Value[Cid]],
+  ): Boolean = {
+    valueIsReplayedBy(recorded.key, replayed.key) && recorded.maintainers == replayed.maintainers
+  }
+
+  private[this] def keys[K](entries: ImmArray[(K, _)]): Iterator[K] =
+    entries.iterator.map(_._1)
+
+  private[this] def values[V](entries: ImmArray[(_, V)]): Iterator[V] =
+    entries.iterator.map(_._2)
+
+  private def valueIsReplayedBy(
+      recorded: Value[Cid],
+      replayed: Value[Cid],
+  ): Boolean = {
+
+    import Value._
+
+    @tailrec
+    def loop(tuples: LazyList[(Value[Cid], Value[Cid])]): Boolean =
+      tuples match {
+        case LazyList.cons(tuple, rest) =>
+          tuple match {
+            case (
+                  ValueEnum(recordedTyCon, recordedName),
+                  ValueEnum(replayedTyCon, replayedName),
+                ) =>
+              recordedTyCon == replayedTyCon &&
+                recordedName == replayedName &&
+                loop(rest)
+            case (recordedLeaf: ValueCidlessLeaf, replayedLeaf: ValueCidlessLeaf) =>
+              recordedLeaf == replayedLeaf &&
+                loop(rest)
+            case (ValueContractId(recordedValue), ValueContractId(replayedValue)) =>
+              recordedValue === replayedValue &&
+                loop(rest)
+            case (
+                  ValueRecord(recordedTyCon, recordedFields),
+                  ValueRecord(replayedTyCon, replayedFields),
+                ) =>
+              recordedTyCon == replayedTyCon &&
+                recordedFields.length == replayedFields.length &&
+                (keys(recordedFields) zip keys(replayedFields)).forall {
+                  case (None, _) => true
+                  case (recorded, replayed) => recorded == replayed
+                } &&
+                loop((values(recordedFields) zip values(recordedFields)) ++: rest)
+            case (
+                  ValueVariant(recordedTyCon, recordedVariant, recordedValue),
+                  ValueVariant(replayedTyCon, replayedVariant, replayedValue),
+                ) =>
+              recordedTyCon == replayedTyCon &&
+                recordedVariant == replayedVariant &&
+                loop((recordedValue, replayedValue) +: rest)
+            case (ValueList(recordedValues), ValueList(replayedValues)) =>
```

Comment:
I know this didn’t change but I don’t understand why we don’t have to check the length here.

Reply:
It is in fact changed (I update the PR description).
We have to check the length (you already catch that in another PR that I try to slice apart)
Thanks.